### PR TITLE
[DISCO-3003] Add Accept-Language header

### PR DESCRIPTION
**Pull request by tiftran**
## References

JIRA: [DISCO-3003](https://mozilla-hub.atlassian.net/browse/DISCO-3003)

## Description
Suggest endpoint to parse the Accept-Language header and filter out languages to only the ones it support

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ ] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [ ] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [ ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[DISCO-3003]: https://mozilla-hub.atlassian.net/browse/DISCO-3003?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: merino/configs/default.toml [hunk 1]
```diff
@@ -42,6 +42,9 @@ disabled_providers = []
 # MERINO_RUNTIME__DEFAULT_SUGGESTIONS_RESPONSE_TTL_SEC
 default_suggestions_response_ttl_sec = 300 # 5 mins
 
+# MERINO_RUNTIME__DEFAULT_LANGUAGES
+default_languages  = ["de", "en-CA", "en-GB", "en-US", "es-ES", "fr", "it", "pl", "pt-BR", "ru", "zh-CN"]
```
Comment (tiftran) [inline merino/configs/default.toml:46]: I thought our "tier 1" locales would be a good start.

File: merino/web/api_v1.py [hunk 2]
```diff
@@ -174,6 +175,7 @@ async def suggest(
         search_from = default_providers
 
     lookups: list[Task] = []
+    _ = get_accepted_languages(accept_language)
```
Comment (tiftran) [inline merino/web/api_v1.py:178]: since it's not being currently used.

Comment (gruberb) [reply merino/web/api_v1.py:178]: Then why calling it? Or at least, add a comment why it's not being used and still  being called here?

File: merino/web/api_v1.py [hunk 3]
```diff
@@ -341,3 +343,22 @@ async def curated_content(
     [curated-topics-doc]: https://mozilla-hub.atlassian.net/wiki/x/LQDaMg
     """
     return await provider.fetch(curated_recommendations_request)
+
+
+def get_accepted_languages(languages: str) -> list[str]:
+    """Retrieve filtered list of languages that merino accepts."""
+    if languages:
```
Comment (ncloudioj) [inline merino/web/api_v1.py:350]: Some notes:
- Let's wrap the parsing block inside a try..except block to protect against malformed inputs
- Let's parse all the accept languages and leave the default_languages filtering to individual downstream consumers
- Optionally, we can decorate this parser with `@functools.lru_cache(maxsize=1000)` for memorization

**Review by ncloudioj — CHANGES_REQUESTED** [review]

File: merino/web/api_v1.py (hunk 2 above)
Comment (tiftran) [reply merino/web/api_v1.py:178]: gah, sorry i didn't finish my thought above. I was planning to have it consumed/used in a follow up PR. Yes agreed, was going to put it into SuggestionRequest, wanted to keep this PR tiny 😅

File: merino/web/api_v1.py (hunk 3 above)
Comment (tiftran) [reply merino/web/api_v1.py:350]: Agreed that we should wrap the block in a try/except block, but not sure what exceptions that could pop up. `split` returns a list even if the split string does not exist. quality logic also handles an index error if the quality value doesn't exist.

File: merino/web/api_v1.py [hunk 4]
```diff
@@ -67,7 +68,8 @@
 )
 async def suggest(
     request: Request,
-    q: str = Query(max_length=QUERY_CHARACTER_MAX),
+    q: Annotated[str, Query(max_length=QUERY_CHARACTER_MAX)],
+    accept_language: Annotated[str | None, Header()] = None,
```
Comment (ncloudioj) [inline merino/web/api_v1.py:72]:
```suggestion
    accept_language: Annotated[str | None, Header(max_length=HEADER_CHARACTER_MAX)] = None,
```

File: merino/web/api_v1.py [hunk 5]
```diff
@@ -341,3 +343,25 @@ async def curated_content(
     [curated-topics-doc]: https://mozilla-hub.atlassian.net/wiki/x/LQDaMg
     """
     return await provider.fetch(curated_recommendations_request)
+
+
+@functools.lru_cache(maxsize=1000)
+def get_accepted_languages(languages: str) -> list[str]:
```
Comment (ncloudioj) [inline merino/web/api_v1.py:349]:
```suggestion
def get_accepted_languages(languages: str | None) -> list[str]:
```

File: tests/unit/web/test_api_v1.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    ("languages", "expected_filtered_languages"),
+    [
+        ("*", ["en-US"]),
+        ("en-US", ["en-US"]),
+        ("en-US,en;q=0.5", ["en-US", "en"]),
+        ("en-US,en;q=0.9,zh-CN;q=0.7", ["en-US", "en", "zh-CN"]),
+        ("en-CA;q=invalid", ["en-US"]),
+    ],
+)
+def test_get_accepted_languages(languages, expected_filtered_languages):
```
Comment (ncloudioj) [inline tests/unit/web/test_api_v1.py:22]: Let's add a test case for the exception handling path.

**Review by ncloudioj — APPROVED** [review]
Looks good, r+wc.

File: tests/unit/web/test_api_v1.py (hunk 6 above)
Comment (tiftran) [reply tests/unit/web/test_api_v1.py:22]: ("en-CA;q=invalid", ["en-US"]), should technically cover the exception path?

Comment (ncloudioj) [reply tests/unit/web/test_api_v1.py:22]: Right, sorry, I didn't realize that.

**Pull request merged**